osu.Game/Screens/Select/SongSelect.cs

```diff
@@ -443,6 +456,34 @@ private void ensurePlayingSelected(bool preview = false)
             }
         }
 
+        private void ensurePlayableRuleset()

```

Comment:
Is there any reason for this to be here rather than in `BeatmapCarousel`? You're duplicating the "can we convert" conditional out of the carousel scope (and doing so twice in the same method). Need to avoid duplicating such code wherever possible.

I'd suggest trying to work this in to `BeatmapCarousel`'s `SelectBeatmap`, which is already called in the necessary code paths.

Then you can probably already check if the beatmap to be selected is currently filtered, and if so, perform the required actions. Also I'm not sure whether we want it to trigger a ruleset change, maybe leave that part out (and call `SelectNextRandom` instead).

Reply:
the ruleset change was to handle the scenario where the next track in the music controller might be an incompatible ruleset. when pressing "next" in music controller loads a random beatmap instead it'd be a bit disconcerting (like in #2055)

Reply:
I think the more common scenario is the one we should cater to initially – entering song select from the main menu after a random beatmap starts playing. We don't want this to switch a user's active ruleset.